elastic/kibana
# [Response Ops][Alerting] SLIs Phase 2 - Counter for on time tasks

Towards https://github.com/elastic/response-ops-team/issues/130

## Summary

This implements the first of 3 SLIs described in https://github.com/elastic/response-ops-team/issues/130 - the metric for keep track of task timeouts. Because this will be used in an SLO, we actually want to keep track of the number of tasks that run "on time" (meaning within the task timeout value). This metric is broken down by task type (and alerting/action tasks are grouped).

## To Verify

Modify a rule type to have a short task timeout and then add a delay within the executor code so it always times out. I did this to the always firing example rule type:

```
diff --git a/x-pack/examples/alerting_example/server/alert_types/always_firing.ts b/x-pack/examples/alerting_example/server/alert_types/always_firing.ts
index 3c048702e47..fc071a38c79 100644
--- a/x-pack/examples/alerting_example/server/alert_types/always_firing.ts
+++ b/x-pack/examples/alerting_example/server/alert_types/always_firing.ts
@@ -56,6 +56,7 @@ export const alertType: RuleType<
   defaultActionGroupId: DEFAULT_ACTION_GROUP,
   minimumLicenseRequired: 'basic',
   isExportable: true,
+  ruleTaskTimeout: '10s',
   async executor({
     services,
     params: { instances = DEFAULT_INSTANCES_TO_GENERATE, thresholds },
@@ -63,6 +64,7 @@ export const alertType: RuleType<
   }) {
     const count = (state.count ?? 0) + 1;

+    await new Promise((r) => setTimeout(r, 20000));
     range(instances)
       .map(() => uuidv4())
       .forEach((id: string) => {
```

Create a rule with this type and let it run. When you see the task cancelled message in the logs, navigate to https://localhost:5601/api/task_manager/metrics?reset=false and you should see that the new metric look something like this: 

You may have to refresh a few times and there is a 30 second reset on these metrics so if you wait too long you may have to run the rule again.

```
{
  "task_run": {
    "timestamp": "2023-09-08T19:59:42.627Z",
    "value": {
      "overall": {
        "success": 4,
        "on_time": 3,
        "total": 4
      },
      "by_type": {
        "alerting": {
          "success": 1,
          "on_time": 0,
          "total": 1
        }
      }
    }
  }
}
```

Note that whether the task ran "on time" is separate than task success as a task (particularly one that doesn't implement cancellation logic) can finish "successfully" even if it did so after the timeout.
